### PR TITLE
Lab5 - Stringer

### LINKED ISSUE #250: Lab5 - Stringer
- [ ] Create an executable go program in directory 05_stringer/CameronSchafer

- [ ] Make the IPAddr type implement fmt.Stringer to print the address as a dotted quad
- [ ] Call **fmt.Println(IPAddr{127, 0, 0, 1})** in **main** to print:
```
127.0.0.1
```

### PR DESCRIPTION
**Pull request by CameronSchafer**
Fixes #250

Review of colleague's PR #355

#### Changes proposed in this PR:

- Add fmt.stringer functionality
- Test fmt.stringer implementation

### CODE REVIEW

**Comment by joshcarp** [comment]
Hey, your latest commit `c25afb` shouldn't be in there; you're meant to have others code in your branch. 
You can do 
`git reset --hard HEAD~1`
then 
`git rebase upstream/master` to update your branch before merging

File: 03_letters/CameronSchafer/letters_test.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+	if expected != actual {
+		t.Errorf("Unexpected output in main()\nexpected: %q\nactual: %q", expected, actual)
+	}
+}
+
+// sortLetters()
+func TestSortLetters(t *testing.T) {
+	// test cases with descriptions.
+	testCases := []struct {
```
Comment (joshcarp) [inline 03_letters/CameronSchafer/letters_test.go]: You can make testCases a global var and do all the tests, (table tests)
Also might be good to have multiple global vars to test a set of edge cases, ex 
`testWhitespaces` 
`testSpecials`

File: 03_letters/CameronSchafer/letters_test.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			resultArr := sortLetters(test.input)
+			assert.Equalf(t, resultArr, test.expectedArr, "Unexpected output in %v\nexpected: %v,\nactual: %v",
+				test.description, test.expectedArr, resultArr)
+		})
+	}
+}
+
+// letters()
+func TestLetters(t *testing.T) {
+	// test cases with descriptions.
```
Comment (joshcarp) [inline 03_letters/CameronSchafer/letters_test.go]: I don't think comments add anything

File: 03_letters/CameronSchafer/letters_test.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+		{description: "sortLetters [97:7, 98:7]", input: map[rune]int{97: 7, 98: 7},
+			expectedArr: []string{"a:7", "b:7"},
+		},
+		{description: "sortLetters [32:2, 97:1]", input: map[rune]int{32: 2, 97: 1},
+			expectedArr: []string{" :2", "a:1"},
+		},
+	}
+
+	for _, test := range testCases {
+		test := test
```
Comment (joshcarp) [inline 03_letters/CameronSchafer/letters_test.go]: self assignment?
`test:= test`

File: 05_stringer/CameronSchafer/main.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package main
```
Comment (joshcarp) [inline 05_stringer/CameronSchafer/main.go:1]: this file should be in a different pull request

**Review by joshcarp — DISMISSED** [review]
good job! 
just a couple of things that need changing; 
- Sometimes adding comments is a bit redundant, so try to leave them out unless you think they're really needed
- Table driven tests are really nice

**Comment by joshcarp** [comment]
I'd say take a look at the golang github wiki on review comments:
https://github.com/golang/go/wiki/CodeReviewComments
and also the wiki on table driven tests:
https://github.com/golang/go/wiki/TableDrivenTests

File: 03_letters/CameronSchafer/letters.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+	for i := range s {
+		tMap[rune(s[i])]++
+	}
+
+	return tMap
+}
+
+func sortLetters(m map[rune]int) []string {
+	tSlice := []int{}
```
Comment (joshcarp) [inline 03_letters/CameronSchafer/letters.go]: the more `go` way of delaring empty slices would be
`var tSlice []int`
or if we're going for optimisation
`tSlice := make([]int, 0, len(m))` this will preallocate only the needed memory to tSlice

File: 05_stringer/CameronSchafer/main_test.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+// main() test
+func TestMainOutput(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	main()
+	expected := "127.0.0.1\n"
+	actual := buf.String()
+
```
Comment (joshcarp) [inline 05_stringer/CameronSchafer/main_test.go:17]: Not exactly sure, but I imagine there might be a little bug; once you set `out = &buf` it is never set back to `os.Stdout`, so any tests after this one won't be able to output in the standard output. 
I guess `out = os.Stdout` would fix it

Comment (joshcarp) [reply 05_stringer/CameronSchafer/main_test.go:17]: Wait no, this only impacts the main function because it's the only place `out` is used

Comment (joshcarp) [reply 05_stringer/CameronSchafer/main_test.go:17]: Nice!

File: 05_stringer/CameronSchafer/main_test.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+		{description: "[204, 180, 10, 255].String()",
+			input:       IPAddr{204, 180, 10, 255},
+			expectedArr: "204.180.10.255",
+		},
+	}
+
+	for _, test := range testCases {
+		test := test
+		// t.Run creates a sub test and runs it like a normal test
+		t.Run(test.description, func(t *testing.T) {
```
Comment (joshcarp) [inline 05_stringer/CameronSchafer/main_test.go]: I like this

File: 05_stringer/CameronSchafer/main.go (hunk 4 above)
Comment (CameronSchafer) [reply 05_stringer/CameronSchafer/main.go:1]: I somehow added lab3 files in my lab5 pull request but that is fixed now.

File: 03_letters/CameronSchafer/letters_test.go (hunk 2 above)
Comment (CameronSchafer) [reply 03_letters/CameronSchafer/letters_test.go]: This file was removed because wrong lab

File: 03_letters/CameronSchafer/letters_test.go (hunk 1 above)
Comment (CameronSchafer) [reply 03_letters/CameronSchafer/letters_test.go]: This file was removed because wrong lab

File: 03_letters/CameronSchafer/letters_test.go (hunk 3 above)
Comment (CameronSchafer) [reply 03_letters/CameronSchafer/letters_test.go]: This file was removed because wrong lab

File: 03_letters/CameronSchafer/letters.go (hunk 5 above)
Comment (CameronSchafer) [reply 03_letters/CameronSchafer/letters.go]: This file was removed because wrong lab

File: 05_stringer/CameronSchafer/main_test.go (hunk 7 above)
Comment (CameronSchafer) [reply 05_stringer/CameronSchafer/main_test.go]: These sub tests are so good 👍

File: 05_stringer/CameronSchafer/main_test.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+		expected, actual)
+}
+
+// String()
+func TestFmtStringer(t *testing.T) {
+	testCases := map[string]struct {
+		input       IPAddr
+		expectedArr string
+	}{
+		"127.0.0.1": {input: IPAddr{127, 0, 0, 1},
```
Comment (brvishnu92) [inline 05_stringer/CameronSchafer/main_test.go:27]: You could test boundary cases like single value, no value, etc.

**Review by brvishnu92 — DISMISSED** [review]
Small nitpicks in test cases to implement boundary scenarios.

File: 05_stringer/CameronSchafer/main_test.go (hunk 8 above)
Comment (CameronSchafer) [reply 05_stringer/CameronSchafer/main_test.go:27]: Done,
Thank you for the hint.

**Review by jessodri — APPROVED** [review]
Looks good to me!

File: 05_stringer/CameronSchafer/main_test.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+func TestFmtStringer(t *testing.T) {
+	testCases := map[string]struct {
+		input       IPAddr
+		expectedArr string
+	}{
+		"127.0.0.1": {input: IPAddr{127, 0, 0, 1},
+			expectedArr: "127.0.0.1"},
+		"192.168.1.1": {input: IPAddr{192, 168, 1, 1},
```
Comment (rokane) [inline 05_stringer/CameronSchafer/main_test.go:29]: I'm not really sure what these two tests are testing `192.168.1.1` and `204.180.10.255`. Just remember to select your tests with a purpose, and look for the possible edge cases and guard around them.

Might have been worthwhile to put a test around the max byte value `255.255.255.255` to ensure that a byte array was used as the underlying type.

**Review by rokane — APPROVED** [review]
Looks good to me, and I am happy to let this through without the extra test I suggested, however in future labs remember to think carefully about what each test scenario is actually achieving 👍
